src/eckit/mpi/Serial.cc

```diff
@@ -192,6 +192,15 @@ Status Serial::wait(Request& req) const {
     }
 }
 
+std::vector<Status> Serial::waitall (std::vector<Request> & rq) const {
+    throw NotImplemented ("Serial::waitall", Here ());
+}
+
+Status Serial::waitany(std::vector<Request> &, int &) const
+{
+    throw NotImplemented ("Serial::waitany", Here ());
+}
+
```

Comment:
Both `Serial::waitall()` and `Serial::waitany()` could be implemented simply by making using of the existing `wait()`.
In this serial implementation of isend/ireceive, internal buffers are being kept alive for each request, and the copy will only happen upon the wait.

Reply:
Hello Willem,

I do not know the Serial class; are you sure that depending on the order we process the request list, this could not lead to a deadlock ?

Philippe

Reply:
Hello Willem,

I have looked at the `SerialRequest` class. Requests of this kind have a `test` method which always returns `true`; this implies that no deadlock can ever happen, but that it also hard to implement decently the `waitany` method : if `waitany` is called multiple times on a vector of requests, then it is not possible to pick a request and decide that is has completed (because all of them have already completed).

Another possibility would be to call wait for all requests and stick to the standard which says that : 

> If all of the requests are MPI_REQUEST_NULL, then indx is returned as MPI_UNDEFINED, and status is returned as an empty status. 

But if we do that, then code such as : 

```
    int count = rqr.size (); 
    while (count > 0)
     {   
       int ireq = -1; 
       mpi::Status st = comm.waitany (rqr, ireq); 
       EXPECT (st.error () == 0); 
       count--;
     }   
```
is guaranteed to deadlock, unless the user tests `ireq == MPI_UNDEFINED`. 

The other issue is that MPI_UNDEFINED is not available in Serial.cc (compiled without MPI).

What should I do here ?

Reply:
I see the problem now, but only for waitany, right?

Would it be sufficient if there was a mechanism to mark requests as completed?

Reply:
Yes, you are right, for waitall, it is sufficient to loop on the list of requests, I did it and it works.

But for waitany, we need to to return a different ireq each time it is called, which means some extra information is required; maybe the solution you propose will work, I do not know the Serial class well enough.

Philippe

Reply:
I can take your branch and figure out a solution to make waitany work in Serial.